File: host/Sources/VibeWireHost/Capture/DisplayCatalog.swift
import Foundation
import ScreenCaptureKit
import CoreGraphics
import AppKit

/// One entry in the DISPLAYS list on 02A–02D and the MON 1 / MON 2 / BOTH
/// segmented control on 03A.
struct DisplayInfo: Sendable, Equatable {
    let id: CGDirectDisplayID
    let name: String
    let width: Int
    let height: Int
    let refreshHz: Int
    let isBuiltIn: Bool
    let isMain: Bool

    var wire: [String: Any] {
        [
            "id": Int(id),
            "name": name,
            "width": width,
            "height": height,
            "hz": refreshHz,
            "isBuiltIn": isBuiltIn,
            "isMain": isMain,
        ]
    }

    /// "Monitor 1 · built-in", "Monitor 2 · Studio"
    static func label(index: Int, isBuiltIn: Bool, modelName: String?) -> String {
        let suffix: String
        if isBuiltIn {
            suffix = "built-in"
        } else if let modelName, !modelName.isEmpty {
            suffix = modelName
        } else {
            suffix = "external"
        }
        return "Monitor \(index) · \(suffix)"
    }
}

/// Enumerates displays and tracks hot-plug. ScreenCaptureKit is the source of
/// truth for what is capturable; CoreGraphics fills in refresh rate and the
/// built-in flag, which SCDisplay does not expose.
actor DisplayCatalog {
    private(set) var displays: [DisplayInfo] = []
    private var lastRefresh: Date?
    /// Once per spell of sleep, not once per 2 s poll.
    private var warnedAsleep = false

    /// Refreshes at most every 2 s unless forced. The home screen polls this
    /// while visible and we do not want to hammer the window server.
    func refresh(force: Bool = false) async -> [DisplayInfo] {
        if !force, let lastRefresh, Date().timeIntervalSince(lastRefresh) < 2 {
            return displays
        }

        do {
            let content = try await SCShareableContent.excludingDesktopWindows(
                false, onScreenWindowsOnly: true
            )
            var result: [DisplayInfo] = []
            for (index, display) in content.displays.enumerated() {
                let id = display.displayID
                let isBuiltIn = CGDisplayIsBuiltin(id) != 0
                let mode = CGDisplayCopyDisplayMode(id)
                let hz = Int((mode?.refreshRate ?? 0).rounded())
                let screen = NSScreen.screens.first {
                    ($0.deviceDescription[NSDeviceDescriptionKey("NSScreenNumber")] as? NSNumber)?
                        .uint32Value == id
                }
                result.append(DisplayInfo(
                    id: id,
                    name: DisplayInfo.label(
                        index: index + 1,
                        isBuiltIn: isBuiltIn,
                        modelName: screen?.localizedName
                    ),
                    width: display.width,
                    height: display.height,
                    // A 0 from CoreGraphics means "unknown", which in practice
                    // is a 60 Hz external panel.
                    refreshHz: hz > 0 ? hz : 60,
                    isBuiltIn: isBuiltIn,
                    isMain: CGDisplayIsMain(id) != 0
                ))
            }
            // An empty answer from ScreenCaptureKit means one of two very
            // different things, and overwriting the list conflated them. With
            // no display online it means the displays are gone, and an empty
            // list is the truth. With displays online but asleep — a locked
            // Mac, one second in — it means the panels are off, and wiping the
            // list is what put "the Mac reports no displays, check Screen
            // Recording permission" on the phone over a permission that was
            // granted. Hold what was last seen; `displaysAsleep` on the status
            // payload is what says the picture is off.
            if result.isEmpty, !displays.isEmpty, SystemServices.displaysAreAsleep() {
                if !warnedAsleep {
                    warnedAsleep = true
                    Log.info(.capture, "displays are asleep; holding the last known list")
                }
                lastRefresh = Date()
                return displays
            }
            warnedAsleep = false
            displays = result
            lastRefresh = Date()
        } catch {
            Log.warn(.capture, "display enumeration failed: \(error)")
        }
        return displays
    }

    func display(id: CGDirectDisplayID) -> DisplayInfo? {
        displays.first { $0.id == id }
    }

    func defaultDisplay() -> DisplayInfo? {
        displays.first(where: \.isMain) ?? displays.first
    }

    /// Screen-recording permission, checked without prompting. The menu bar
    /// uses this to explain why the picture is missing rather than showing a
    /// black rectangle.
    nonisolated static func hasScreenRecordingPermission() -> Bool {
        CGPreflightScreenCaptureAccess()
    }

    nonisolated static func requestScreenRecordingPermission() {
        CGRequestScreenCaptureAccess()
    }
}

File: host-electron/src/main/capture/displayCatalog.ts
import { desktopCapturer, screen, type DesktopCapturerSource, type Display } from 'electron'
import { Log, describeError } from '../core/log'
import type { DisplayInfo } from './captureHost'

/**
 * Enumerates displays and tracks hot-plug. `screen` is the source of truth for
 * geometry and refresh rate; `desktopCapturer` names the source Chromium will
 * capture, matched back to a display by its id.
 *
 * Sizes on the wire are physical pixels, as the Mac host reports them: the
 * picture the phone sees is the captured frame, and the injector clamps in the
 * same space, so a 150 % display is not described as smaller than it is.
 */
export class DisplayCatalog {
  private displays: DisplayInfo[] = []
  private lastRefresh = 0
  private stale = true

  constructor() {
    const changed = (event: string) => () => {
      this.stale = true
      Log.info('capture', `displays changed (${event})`)
    }
    screen.on('display-added', changed('display-added'))
    screen.on('display-removed', changed('display-removed'))
    screen.on('display-metrics-changed', changed('display-metrics-changed'))
  }

  /** Refreshes at most every 2 s unless forced or a display came or went. */
  async refresh(force = false): Promise<DisplayInfo[]> {
    if (!force && !this.stale && Date.now() - this.lastRefresh < 2000) return this.displays
    const primary = screen.getPrimaryDisplay().id
    const result: DisplayInfo[] = screen.getAllDisplays().map((display, index) => ({
      id: display.id,
      name: label(index + 1, display),
      width: Math.round(display.size.width * display.scaleFactor),
      height: Math.round(display.size.height * display.scaleFactor),
      // A 0 means "unknown", which in practice is a 60 Hz external panel.
      hz: display.displayFrequency > 0 ? Math.round(display.displayFrequency) : 60,
      isBuiltIn: display.internal,
      isMain: display.id === primary,
    }))
    if (result.length === 0 && this.displays.length > 0) {
      // Displays online but reported absent means the panels are off, not
      // gone. Hold what was last seen rather than telling the phone the host
      // has no screens.
      Log.info('capture', 'no displays reported; holding the last known list')
    } else {
      this.displays = result
    }
    this.lastRefresh = Date.now()
    this.stale = false
    return this.displays
  }

  display(id: number): DisplayInfo | null {
    return this.displays.find((display) => display.id === id) ?? null
  }

  defaultDisplay(): DisplayInfo | null {
    return this.displays.find((display) => display.isMain) ?? this.displays[0] ?? null
  }

  /** The `desktopCapturer` source for a display, matched by id first and by
   *  position in the list second, saying which strategy won. */
  async source(displayId: number): Promise<DesktopCapturerSource | null> {
    let sources: DesktopCapturerSource[]
    try {
      sources = await desktopCapturer.getSources({ types: ['screen'], thumbnailSize: { width: 0, height: 0 } })
    } catch (error) {
      Log.error('capture', `desktopCapturer failed: ${describeError(error)}`)
      return null
    }
    Log.debug(
      'capture',
      `capture sources: ${sources.map((source) => `${source.id} display_id=${JSON.stringify(source.display_id)} "${source.name}"`).join(' | ') || 'none'}; screens: ${screen
        .getAllDisplays()
        .map((display) => display.id)
        .join(',')}`,
    )
    const byId = sources.find((source) => source.display_id === String(displayId))
    if (byId) return byId
    const index = screen.getAllDisplays().findIndex((display) => display.id === displayId)
    const byOrder = index >= 0 ? sources[index] : undefined
    if (byOrder) {
      Log.warn('capture', `display ${displayId} matched a capture source by position, not by id`)
      return byOrder
    }
    return null
  }
}

/** "Monitor 1 · built-in", "Monitor 2 · DELL U2723QE" */
function label(index: number, display: Display): string {
  const suffix = display.internal ? 'built-in' : display.label?.trim() || 'external'
  return `Monitor ${index} · ${suffix}`
}
